### PR TITLE
Allow to offload / customize key signing operations when using Boring…

### PR DESCRIPTION
**Pull request by normanmaurer**
…SSL.

Motivation:

BoringSSL allows to customize the way how key signing is done an even offload it from the IO thread. We should provide a way to plugin an own implementation when BoringSSL is used.

Modifications:

- Introduce OpenSslPrivateKeyMethod that can be used by the user to implement custom signing by using ReferenceCountedOpenSslContext.setPrivateKeyMethod(...)
- Introduce static methods to OpenSslKeyManagerFactory which allows to create a KeyManagerFactory which supports to do keyless operations by let the use handle everything in OpenSslPrivateKeyMethod.
- Add testcase which verifies that everything works as expected

Result:

A user is able to customize the way how keys are signed.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This depends on https://github.com/netty/netty-tcnative/pull/446

File: handler/src/main/java/io/netty/handler/ssl/OpenSslPrivateKeyMethod.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    int SSL_SIGN_ECDSA_SECP384R1_SHA384 = SSLPrivateKeyMethod.SSL_SIGN_ECDSA_SECP384R1_SHA384;
+    int SSL_SIGN_ECDSA_SECP521R1_SHA512 = SSLPrivateKeyMethod.SSL_SIGN_ECDSA_SECP521R1_SHA512;
+    int SSL_SIGN_RSA_PSS_RSAE_SHA256 = SSLPrivateKeyMethod.SSL_SIGN_RSA_PSS_RSAE_SHA256;
+    int SSL_SIGN_RSA_PSS_RSAE_SHA384 = SSLPrivateKeyMethod.SSL_SIGN_RSA_PSS_RSAE_SHA384;
+    int SSL_SIGN_RSA_PSS_RSAE_SHA512 = SSLPrivateKeyMethod.SSL_SIGN_RSA_PSS_RSAE_SHA512;
+    int SSL_SIGN_ED25519 = SSLPrivateKeyMethod.SSL_SIGN_ED25519;
+    int SSL_SIGN_RSA_PKCS1_MD5_SHA1 = SSLPrivateKeyMethod.SSL_SIGN_RSA_PKCS1_MD5_SHA1;
+
+    /**
+     * Sign the input with given key and return the signed bytes.
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslPrivateKeyMethod.java]: - Sign -> Signs
- with given -> with the given
- return -> returns

File: handler/src/main/java/io/netty/handler/ssl/OpenSslPrivateKeyMethod.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+     * @param signatureAlgorithm    the algorithm to use for signing
+     * @param input                 the digest itself
+     * @param key                   the key if configured, or {@code null} if the key needs to be obtained externally.
+     * @return                      the signed data
+     * @throws Exception
+     */
+    byte[] sign(SSLEngine engine, int signatureAlgorithm, byte[] input, byte[] key)
+            throws Exception;
+    /**
+     * Decrypt the input with the given key and return the decrypted bytes.
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslPrivateKeyMethod.java]: - Decrypt -> Decrypts
- return -> returns

File: handler/src/main/java/io/netty/handler/ssl/OpenSslX509KeyManagerFactory.java [hunk 3]
```diff
@@ -252,31 +252,69 @@ public static OpenSslX509KeyManagerFactory newEngineBased(File certificateChain,
...
+     * See {@link OpenSslX509KeyManagerFactory#newEngineBased(X509Certificate[], String)}.
+     */
+    public static OpenSslX509KeyManagerFactory newKeyless(InputStream chain)
+            throws CertificateException, IOException,
+            KeyStoreException, NoSuchAlgorithmException, UnrecoverableKeyException {
+        return newKeyless(SslContext.toX509Certificates(chain));
+    }
+
+    /**
+     * Create a new initialized {@link OpenSslX509KeyManagerFactory} which will provide its private key by using the
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslX509KeyManagerFactory.java]: - Create -> Creates (or Returns a newly initialized ...?)
- by using the -> by using a .. ? (I'm not a native speaker, so someone please tell us which sounds better.)

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java [hunk 4]
```diff
@@ -517,6 +520,18 @@ public final long sslCtxPointer() {
         }
     }
 
+    @UnstableApi
+    public void setPrivateKeyMethod(OpenSslPrivateKeyMethod method) {
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java]: Missing Javadoc?

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java [hunk 5]
```diff
@@ -884,4 +899,33 @@ static OpenSslKeyMaterialProvider providerFor(KeyManagerFactory factory, String
...
+        private final OpenSslPrivateKeyMethod keyMethod;
+        PrivateKeyMethod(OpenSslEngineMap engineMap, OpenSslPrivateKeyMethod keyMethod) {
+            this.engineMap = engineMap;
+            this.keyMethod = keyMethod;
+        }
+
+        private ReferenceCountedOpenSslEngine retrieveEngine(long ssl) throws SSLException {
+            ReferenceCountedOpenSslEngine engine = engineMap.get(ssl);
+            if (engine == null) {
+                throw new SSLException("Could not find Engine");
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java]: nit: Could not find an OpenSslEngine ?

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 6]
```diff
@@ -889,7 +889,8 @@ public final SSLEngineResult wrap(
                             // to write encrypted data to. This is an OVERFLOW condition.
                             // [1] https://www.openssl.org/docs/manmaster/ssl/SSL_write.html
                             return newResult(BUFFER_OVERFLOW, status, bytesConsumed, bytesProduced);
-                        } else if (sslError == SSL.SSL_ERROR_WANT_X509_LOOKUP) {
+                        } else if (sslError == SSL.SSL_ERROR_WANT_X509_LOOKUP
+                                || sslError == SSL.SSL_ERROR_WANT_PRIVATE_KEY_OPERATION) {
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: nit: I see you are mixing two styles:

```java
if (a ||
    b) { ... }

if (a
    || b) { ... }
```

IIRC we preferred the former since the beginning of the project. ;-)

File: handler/src/test/java/io/netty/handler/ssl/OpenSslPrivateKeyMethodTest.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+        CERT = new SelfSignedCertificate();
+    }
+
+    @AfterClass
+    public static void destroy() {
+        GROUP.shutdownGracefully();
+        CERT.delete();
+    }
+
+    private final String rfcCipherName = "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256";
```
Comment (trustin) [inline handler/src/test/java/io/netty/handler/ssl/OpenSslPrivateKeyMethodTest.java]: Could be static and renamed to `RFC_CIPHER_NAME`.

File: handler/src/test/java/io/netty/handler/ssl/OpenSslPrivateKeyMethodTest.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+        ((OpenSslContext) sslServerContext).setPrivateKeyMethod(new OpenSslPrivateKeyMethod() {
+            @Override
+            public byte[] sign(SSLEngine engine, int signatureAlgorithm, byte[] input, byte[] key) throws Exception {
+                signCalled.set(true);
+                Assert.assertEquals(OpenSslPrivateKeyMethod.SSL_SIGN_RSA_PKCS1_SHA256, signatureAlgorithm);
+
+                if (delegate && OpenSslContext.USE_TASKS) {
+                    Assert.assertEquals(DelegateThread.class, Thread.currentThread().getClass());
+                } else {
+                    Assert.assertNotEquals(DelegateThread.class, Thread.currentThread().getClass());
```
Comment (trustin) [inline handler/src/test/java/io/netty/handler/ssl/OpenSslPrivateKeyMethodTest.java]: Consider using static imports for all `Assert` calls.

File: pom.xml [hunk 9]
```diff
@@ -282,8 +282,8 @@
     <!-- Configure the os-maven-plugin extension to expand the classifier on                  -->
     <!-- Fedora-"like" systems. This is currently only used for the netty-tcnative dependency -->
     <os.detection.classifierWithLikes>fedora</os.detection.classifierWithLikes>
-    <tcnative.artifactId>netty-tcnative</tcnative.artifactId>
-    <tcnative.version>2.0.23.Final</tcnative.version>
+    <tcnative.artifactId>netty-tcnative-boringssl-static</tcnative.artifactId>
```
Comment (normanmaurer) [inline pom.xml]: I will revert this one before merging

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 10]
```diff
@@ -966,7 +968,12 @@ private SSLException shutdownWithError(String operation, int sslError, int error
         if (handshakeState == HandshakeState.FINISHED) {
             return new SSLException(errorString);
         }
-        return new SSLHandshakeException(errorString);
+        SSLHandshakeException exception = new SSLHandshakeException(errorString);
+        if (handshakeException != null) {
+            exception.initCause(handshakeException);
+            handshakeException = null;
+        }
+        return exception;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: As this change is not related to this PR I created https://github.com/netty/netty/pull/9008.

**Comment by normanmaurer** [comment]
I will merge this now as the code is really isolated and so will not affect other things. This will allow me to make progress
